webapp/src/ts/modals/training-cards/training-cards.component.ts

```diff
@@ -0,0 +1,217 @@
+      this.form = await this.enketoService.render(selector, form, undefined, undefined, this.resetFormError.bind(this));
+      this.loadingContent = false;
+      this.recordTelemetryPostRender();
+    } catch(error) {
+      this.setError();
+      console.error('Error rendering form.', error);
+    }
+  }
+
+  private subscribeToStore() {
+    const reduxSubscription = combineLatest(
+      this.store.select(Selectors.getEnketoStatus),
+      this.store.select(Selectors.getEnketoSavingStatus),
+      this.store.select(Selectors.getEnketoError),
+      this.store.select(Selectors.getTrainingCard),
+    ).subscribe(([
+      enketoStatus,
+      enketoSaving,
+      enketoError,
+      trainingForm,
+    ]) => {
+      this.enketoStatus = enketoStatus;
+      this.enketoSaving = enketoSaving;
+      this.enketoError = enketoError;
+
+      if (trainingForm && trainingForm !== this.trainingForm) {
+        this.trainingForm = trainingForm;
+        this.loadForm();
+      }
+    });
+    this.subscription.add(reduxSubscription);
+  }
+
+  private reset() {
+    this.resetFormError();
+    this.contentError = false;
+    this.trainingForm = null;
+    this.form = null;
+    this.loadingContent = true;
+    this.hideModalFooter = true;
+  }
+
+  private resetFormError() {
+    if (this.enketoError) {
+      this.globalActions.setEnketoError(null);
+    }
+  }
+
+  setError() {
+    this.errorTranslationKey = 'training_cards.error.loading';
+    this.loadingContent = false;
+    this.hideModalFooter = false;
+    this.contentError = true;
+  }
+
+  async saveForm() {
+    if (this.enketoSaving) {
+      console.debug('Attempted to call TrainingCardsComponent:saveForm more than once');
+      return;
+    }
+
+    this.recordTelemetryPreSave();
+    this.globalActions.setEnketoSavingStatus(true);
+    this.resetFormError();
+
+    try {
+      const docs = await this.enketoService.save(this.trainingForm, this.form, this.geoHandle);
+      console.debug('Saved form and associated docs', docs);
+      this.globalActions.setSnackbarContent(this.translateService.instant('training_cards.form.saved'));
+      this.globalActions.setEnketoSavingStatus(false);
+      this.enketoService.unload(this.form);
+      this.recordTelemetryPostSave();
+      this.close();
+
+    } catch(error) {
+      this.globalActions.setEnketoSavingStatus(false);
+      console.error('Error submitting form data: ', error);
+      const message = await this.translateService.get('training_cards.error.save');
+      this.globalActions.setEnketoError(message);
+    }
+  }
+
+  private recordTelemetryPostRender() {
+    this.telemetryData.postRender = Date.now();
+    this.telemetryData.action = 'add';
+    this.telemetryData.form = this.trainingForm.replace('training:', '');
+
+    this.telemetryService.record(
+      `enketo:training:${this.telemetryData.form}:${this.telemetryData.action}:render`,
+      this.telemetryData.postRender - this.telemetryData.preRender
+    );
```

Comment:
I don't think this is a good idea. For one thing it'll make it hard to find the form because you will need to know it's a training form and add the prefix, but also it may clash with another form with the same name. The telemetry key is different but you're already adding `training:` back there so I think you can just remove it, eg:

```suggestion
    this.telemetryData.form = this.trainingForm;

    this.telemetryService.record(
      `enketo:${this.telemetryData.form}:${this.telemetryData.action}:render`,
      this.telemetryData.postRender - this.telemetryData.preRender
    );
```

This goes for the other telemetry items below too.

Reply:
I added the `enketo:training:` piece to keep consistency with what we do with [reports forms](https://github.com/medic/cht-core/blob/master/webapp/src/ts/modules/reports/reports-add.component.ts#L219) and [contact forms](https://github.com/medic/cht-core/blob/1f0f990150bf2e7df0242185f87365e96dbbecf9/webapp/src/ts/modules/contacts/contacts-report.component.ts#L154). 
Just to double check, do we want to keep consistency?

Reply:
Keep the `taining:` bit in the ID, just don't remove it from the `form` property. My suggestion is consistent with those links you sent.